### ISSUE TITLE
TerserPlugin ignores sourceMap setting when using TypeScript loader. No sourcemap is generated.

### ISSUE THREAD

**Issue by versedi** [symfony/webpack-encore#637]
When using Encore + Typescript + TerserPlugin when you enable the TerserPlugin with following settings:

```
    Encore.addPlugin(
        new TerserPlugin({
            terserOptions: {
                sourceMap: true,
                cache: !Encore.isProduction(),
                parallel: true,
                output: {
                    // comments: false,
                },
            },
        }),
    );
```

### The source maps for JS pointing to TS files are no longer generated at all. 

If I wrap the `addPlugin` into `Encore.isProduction()` the sourcemaps are generated fine for development build:

```
if(Encore.isProduction()) {
    Encore.addPlugin(
        new TerserPlugin({
            terserOptions: {
                sourceMap: true,
                cache: !Encore.isProduction(),
                parallel: true,
                output: {
                    // comments: false,
                },
            },
        }),
    );
}
```

The `sourceMap` option value `true` or `false` doesn't affect the build at all.

1. Is there anyway to generate sourceMaps for JS files while compiling typescript and minimizing it? 
2. Could using a custom minifier in TerserPlugin solve this issue?
3. Does it work for anyone else or is this widespread issue?

**Comment by stof** [comment]
Could it be an issue that you have 2 terser plugins running in production (as Encore already applies minification in prod mode) ?
Btw, you should really no add a terser plugin in dev. The minification would slow down your work a lot (in my experience, half of the build time is spent in the minifier)

**Comment by versedi** [comment]
> Could it be an issue that you have 2 terser plugins running in production (as Encore already applies minification in prod mode) ?

After clearing all node_modules, removing TerserPlugin from the build completely and making sure `Encore.enableSourceMaps(true);` is enabled for both environments - I managed to finally find the culprit. 

I've had the `Encore.enableSourceMaps(true)`  after loading the config for check if it's dev environment and configuring the dev-server - it was obviously ommited.

---

>  Btw, you should really no add a terser plugin in dev. The minification would slow down your work a lot (in my experience, half of the build time is spent in the minifier)

I've only unwrapped it from the `if(Encore.isProduction())` cause I had troubles with generating sourceMaps for Sentry in production while it worked in development.  Sorry for describing it the other way - it indeed sounded how you understood it. 

I'm not sure what was the issue that we had to enable TerserPlugin so I'm gonna go through the log and find what it was - if it the issue still persists I'm gonna fill a new issue cause this one will be unrelated. But it might just be an old version that I've started the development on or some other dependency issue. I hope it is :) 

If I recall correctly I had troubles with minification being applied properly in production mode - that's why I've loaded the Terser Plugin. It seems to be gone for now. 



For future Googlers with same trouble: 
# Don't use TerserPlugin with Encore - it already has it's own minimization and sourcemaps configuration. 

## If you use `const config = Encore.getWebpackConfig();` make sure everything done through `Encore.function(param)` is done before that - it doesn't work after expanding the webpack config.

**Comment by versedi** [comment]
> Could it be an issue that you have 2 terser plugins running in production (as Encore already applies minification in prod mode) ?

I'm not sure why I've enabled it in the first place but it could have been an issue with minification or polyfills. 

>  Btw, you should really no add a terser plugin in dev. The minification would slow down your work a lot (in my experience, half of the build time is spent in the minifier)

I've badly worded the issue, I've only unwrapped it cause it didn't worked for production, sorry about being not clear. 


The issue is resolved now but I'm gonna need to run some more tests tommorow and go through the log why it was done  - if the issue persists it will be unrelated though so I'm gonna close this one for now. 

It was two mistakes combined, for future seekers:

- Enabling the *Terser Plugin* with *Webpack Encore* in the first place - see above. 

- Enabling source maps by `Encore.enableSourceMaps(true);`  **after loading the advanced config** `const config = Encore.getWebpackConfig();` for configuring dev-server. 

That can't work. 

I did it cause I didn't needed sourcemaps on production earlier but now that I was implementing the SentryWebpackPlugin I somehow skipped that part. 

The sourceMaps for production will be generated but not only sent to Sentry and removed before deploy. 

Thank you for pointing me in the right direction about the default minimization. 

Did it changed recently or is it possible the docs are outdated in that term?
